enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,424 @@
+
+### API Extensions
+
+N/A - no api extensions are being introduced in this proposal
+
+### Implementation Details/Notes/Constraints [optional]
+
+The CVO already has the ability to respect annotations on resources, as can be seen
+[here](https://github.com/openshift/cluster-kube-apiserver-operator/blob/c03c9edf5fddf4e3fb1bc6d7afcd2a2284ca03d8/manifests/0000_20_kube-apiserver-operator_06_deployment.yaml#L10) and leveraged [here](https://github.com/openshift/hypershift/blob/b98bc43467a73921891ede7d6477757b647f1985/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L152-L153).
+This proposal consists of two parts:
+
+1) Formalizing a concept of a "capability" annotation which allows a given resource to be excluded based
+on installer input. For example the console related resources could be annotated as
+
+```yaml
+annotations:
+  capability.openshift.io/console: "true"
+```
+
+Note: "true" means "included by default unless explicitly excluded", in the future we may enable capabilities
+to mark themselves as "false" meaning "excluded by default, unless explicitly included".
+
+
+2) Defining an [install config api](https://github.com/openshift/installer/blob/790048166067273d34f76bea4220fa395b1cce1b/pkg/types/installconfig.go#L70) field whereby the user can opt out of specific capabilities.
+
+```yaml
+InstallConfig.Capabilities
+  Default: [include|exclude]
+  Exclude:
+  - console
+  - samples
+  Include:
+  - registry
+```
+
+Which resources ultimately get installed for a given cluster would be the set of resources encompassed
+by the CLUSTER_PROFILE(if any), minus any resources explicitly excluded by the excluded capabilities configuration,
+plus any explicitly included capabilities.  Any resource that is not included by the CLUSTER_PROFILE, is not
+included via the Include stanza, and is not Excluded via the Exclude stanza, will be filtered or not based on
+the `Default` value (include or exclude).
+
+The `Default` field (carried through to the CVO) allows an Admin to choose what behavior they want for new
+capabilities that might come as part of a future upgrade (either have the new capability installed during
+the upgrade, or exclude it)
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.
+
+
+Examples of candidate components to be treated as capabilities:
+
+* console
+* imageregistry
+* samples
+* cluster baremetal operator
+* olm/marketplace
+* csi-*
+* insights
+* monitoring
+* ???
+
+
+
+3) Pass the list of filtered annotations to the CVO.  This is distinct from overrides because overrides
+put the cluster in an unsupported state.  Filtered annotations are supported for upgrades.  The filtered
+components will be listed in the [ClusterVersion](https://github.com/openshift/api/blob/4436dc8be01e8dcd8b250e1b32bb0fbd64ba78ac/config/v1/types_cluster_version.go#L35) object:
+
+```yaml
+  spec:
+    capabilities:
+      default: [include|exclude]
```

Comment:
How will the CVO recognize new capabilities after an upgrade? I assume via some status field(s), but do we have to worry about any race conditions in updating that as part of the upgrade process? There was discussion elsewhere of building an artifact as part of the payload to list all of the capabilities. Would that help, since it would mean there's a list of the names all in 1 place?

Reply:
a resource that has a capability annotation, is a capability.  If the resource doesn't have the annotation, the CVO applies it.  If the resource does have the annotation, then it needs to look that this spec config to determine what to do.